File: flyweight/flyweight.py
class Icon:
    def __init__(self, icon_type):
        self.icon_type = icon_type
        self.attribute = "original"

class IconFactory:
    _icons = {}

    @classmethod
    def get_icon(cls, icon_type):
        if icon_type not in cls._icons:
            cls._icons[icon_type] = Icon(icon_type)
        return cls._icons[icon_type]

# Client code
icon1 = IconFactory.get_icon("PDF")
icon2 = IconFactory.get_icon("PDF")

print(icon1.attribute)  # Output: "original"
print(icon2.attribute)  # Output: "original"

icon1.attribute = "changed"

print(icon1.attribute)  # Output: "changed"
print(icon2.attribute)  # Output: "changed"

File: flyweight/flyweight.java
public class Main {
    public static void main(String[] args) {
        Icon icon1 = IconFactory.getIcon("PDF");
        Icon icon2 = IconFactory.getIcon("PDF");

        System.out.println(icon1.getAttribute());  // Output: "original"
        System.out.println(icon2.getAttribute());  // Output: "original"
    }
}

class Icon {
    private String iconType;
    private String attribute;

    public Icon(String iconType) {
        this.iconType = iconType;
        this.attribute = "original";
    }

    public String getAttribute() {
        return attribute;
    }

    public void setAttribute(String attribute) {
        this.attribute = attribute;
    }
}

class IconFactory {
    private static final java.util.HashMap<String, Icon> icons = new java.util.HashMap<>();

    public static Icon getIcon(String iconType) {
        if (!icons.containsKey(iconType)) {
            icons.put(iconType, new Icon(iconType));
        }
        return icons.get(iconType);
    }
}
